File: examples/javascript/node-cjs/src/advanced_usage.ts
import type {
  TransliterationRecipe,
  TransliteratorConfig,
} from "@yosina-lib/yosina";
import { makeTransliterator } from "@yosina-lib/yosina";

const processTextSamples = (samples: [string, string][]) => {
  console.log("Processing text samples with character codes:");
  for (const [text, description] of samples) {
    console.log(`\n${description}:`);
    console.log(`  Original: '${text}'`);

    // Show character codes for clarity
    const charCodes = text
      .split("")
      .map(
        (c) =>
          `U+${c.codePointAt(0)?.toString(16).toUpperCase().padStart(4, "0")}`,
      )
      .join(" ");
    console.log(`  Codes:    ${charCodes}`);
  }
};

const main = async () => {
  console.log("=== Advanced Yosina JavaScript Usage Examples ===\n");

  // 1. Web scraping text normalization
  {
    console.log("1. Web Scraping Text Normalization");
    console.log("   (Typical use case: cleaning scraped Japanese web content)");

    const webScrapingRecipe: TransliterationRecipe = {
      kanjiOldNew: true,
      replaceSpaces: true,
      replaceSuspiciousHyphensToProlongedSoundMarks: true,
      replaceIdeographicAnnotations: true,
      replaceRadicals: true,
      combineDecomposedHiraganasAndKatakanas: true,
    };

    const normalizer = await makeTransliterator(webScrapingRecipe);

    // Simulate messy web content
    const messyContent: [string, string][] = [
      [
        "これは　テスト です。",
        "Mixed spaces from different sources",
      ],
      ["コンピューター-プログラム", "Suspicious hyphens in katakana"],
      ["古い漢字：廣島、檜、國", "Old-style kanji forms"],
      ["部首：⺅⻌⽊", "CJK radicals instead of regular kanji"],
    ];

    for (const [text, description] of messyContent) {
      const cleaned = normalizer(text);
      console.log(`  ${description}:`);
      console.log(`    Before: '${text}'`);
      console.log(`    After:  '${cleaned}'`);
      console.log();
    }
  }

  // 2. Document standardization
  {
    console.log("2. Document Standardization");
    console.log("   (Use case: preparing documents for consistent formatting)");

    const documentRecipe: TransliterationRecipe = {
      toFullwidth: true,
      replaceSpaces: true,
      kanjiOldNew: true,
      combineDecomposedHiraganasAndKatakanas: true,
    };

    const documentStandardizer = await makeTransliterator(documentRecipe);

    const documentSamples: [string, string][] = [
      ["Hello World 123", "ASCII text to full-width"],
      ["カ゛", "Decomposed katakana with combining mark"],
      ["檜原村", "Old kanji in place names"],
    ];

    for (const [text, description] of documentSamples) {
      const standardized = documentStandardizer(text);
      console.log(`  ${description}:`);
      console.log(`    Input:  '${text}'`);
      console.log(`    Output: '${standardized}'`);
      console.log();
    }
  }

  // 3. Search index preparation
  {
    console.log("3. Search Index Preparation");
    console.log("   (Use case: normalizing text for search engines)");

    const searchRecipe: TransliterationRecipe = {
      kanjiOldNew: true,
      replaceSpaces: true,
      toHalfwidth: true,
      replaceSuspiciousHyphensToProlongedSoundMarks: true,
    };

    const searchNormalizer = await makeTransliterator(searchRecipe);

    const searchSamples: [string, string][] = [
      ["東京スカイツリー", "Famous landmark name"],
      ["プログラミング言語", "Technical terms"],
      ["コンピューター-サイエンス", "Academic field with suspicious hyphen"],
    ];

    for (const [text, description] of searchSamples) {
      const normalized = searchNormalizer(text);
      console.log(`  ${description}:`);
      console.log(`    Original:   '${text}'`);
      console.log(`    Normalized: '${normalized}'`);
      console.log();
    }
  }

  // 4. Custom pipeline example
  {
    console.log("4. Custom Processing Pipeline");
    console.log("   (Use case: step-by-step text transformation)");

    // Create multiple transliterators for pipeline processing
    const steps: [string, TransliteratorConfig[]][] = [
      ["Spaces", [["spaces", {}]]],
      [
        "Old Kanji",
        [
          ["ivs-svs-base", { mode: "ivs-or-svs", charset: "unijis_2004" }],
          ["kanji-old-new", {}],
          ["ivs-svs-base", { mode: "base", charset: "unijis_2004" }],
        ],
      ],
      ["Width", [["jisx0201-and-alike", { u005cAsYenSign: false }]]],
      ["ProlongedSoundMarks", [["prolonged-sound-marks", {}]]],
    ];

    const transliterators: [string, (text: string) => string][] = [];
    for (const [name, config] of steps) {
      const transliterator = await makeTransliterator(config);
      transliterators.push([name, transliterator]);
    }

    const pipelineText = "hello\u{3000}world ﾃｽﾄ 檜-システム";
    let currentText = pipelineText;

    console.log(`  Starting text: '${currentText}'`);

    for (const [stepName, transliterator] of transliterators) {
      const previousText = currentText;
      currentText = transliterator(currentText);
      if (previousText !== currentText) {
        console.log(`  After ${stepName}: '${currentText}'`);
      }
    }

    console.log(`  Final result: '${currentText}'`);
  }

  // 5. Unicode normalization showcase
  {
    console.log("5. Unicode Normalization Showcase");
    console.log("   (Demonstrating various Unicode edge cases)");

    const unicodeRecipe: TransliterationRecipe = {
      replaceSpaces: true,
      replaceMathematicalAlphanumerics: true,
      replaceRadicals: true,
    };

    const unicodeNormalizer = await makeTransliterator(unicodeRecipe);

    const unicodeSamples: [string, string][] = [
      ["\u{2003}\u{2002}\u{2000}", "Various em/en spaces"],
      ["\u{3000}\u{00A0}\u{202F}", "Ideographic and narrow spaces"],
      ["⺅⻌⽊", "CJK radicals"],
      ["\u{1D400}\u{1D401}\u{1D402}", "Mathematical bold letters"],
    ];

    processTextSamples(unicodeSamples);

    for (const [text, description] of unicodeSamples) {
      const normalized = unicodeNormalizer(text);
      console.log(`  Result:   '${normalized}'`);
      console.log();
    }
  }

  // 6. Performance considerations
  {
    console.log("6. Performance Considerations");
    console.log("   (Reusing transliterators for better performance)");

    const perfRecipe: TransliterationRecipe = {
      kanjiOldNew: true,
      replaceSpaces: true,
    };

    const perfTransliterator = await makeTransliterator(perfRecipe);

    // Simulate processing multiple texts
    const texts = [
      "これはテストです",
      "檜原村は美しい",
      "hello　world",
      "プログラミング",
    ];

    console.log(
      `  Processing ${texts.length} texts with the same transliterator:`,
    );
    for (let i = 0; i < texts.length; i++) {
      const text = texts[i];
      const result = perfTransliterator(text);
      console.log(`    ${i + 1}: '${text}' → '${result}'`);
    }
  }

  console.log("\n=== Advanced Examples Complete ===");
};

main().catch(console.error);

File: rust/examples/advanced_usage.rs
//! Advanced usage example for Yosina Rust library.
//! This example demonstrates complex text processing scenarios.

use anyhow::Result;
use yosina::transliterators::{
    Charset, IvsSvsBaseMode, IvsSvsBaseTransliteratorOptions, Jisx0201AndAlikeTransliteratorOptions,
};
use yosina::{make_transliterator, TransliterationRecipe, TransliteratorConfig};

fn process_text_samples(samples: &[(String, &str)]) {
    println!("Processing text samples with character codes:");
    for (text, description) in samples {
        println!("\n{description}:");
        println!("  Original: '{text}'");

        // Show character codes for clarity
        let char_codes: Vec<String> = text
            .chars()
            .map(|c| format!("U+{:04X}", c as u32))
            .collect();
        println!("  Codes:    {}", char_codes.join(" "));
    }
}

fn main() -> Result<()> {
    println!("=== Advanced Yosina Rust Usage Examples ===\n");

    // 1. Web scraping text normalization
    println!("1. Web Scraping Text Normalization");
    println!("   (Typical use case: cleaning scraped Japanese web content)");

    let web_scraping_recipe = TransliterationRecipe {
        kanji_old_new: true,
        replace_spaces: true,
        replace_suspicious_hyphens_to_prolonged_sound_marks: true,
        replace_ideographic_annotations: true,
        replace_radicals: true,
        combine_decomposed_hiraganas_and_katakanas: true,
        ..Default::default()
    };

    let web_normalizer = make_transliterator(&web_scraping_recipe)?;

    // Simulate messy web content
    let messy_content = vec![
        (
            "これは\u{3000}テスト\u{00A0}です。".to_string(),
            "Mixed spaces from different sources",
        ),
        (
            "コンピューター-プログラム".to_string(),
            "Suspicious hyphens in katakana",
        ),
        (
            "古い漢字：廣島、檜、國".to_string(),
            "Old-style kanji forms",
        ),
        (
            "部首：⺅⻌⽊".to_string(),
            "CJK radicals instead of regular kanji",
        ),
    ];

    for (text, description) in &messy_content {
        let cleaned = web_normalizer(text)?;
        println!("  {description}:");
        println!("    Before: '{text}'");
        println!("    After:  '{cleaned}'");
        println!();
    }

    // 2. Document standardization
    println!("2. Document Standardization");
    println!("   (Use case: preparing documents for consistent formatting)");

    let document_recipe = TransliterationRecipe {
        to_fullwidth: yosina::recipes::ToFullWidthOptions::Yes {
            u005c_as_yen_sign: false,
        },
        replace_spaces: true,
        kanji_old_new: true,
        combine_decomposed_hiraganas_and_katakanas: true,
        ..Default::default()
    };

    let document_standardizer = make_transliterator(&document_recipe)?;

    let document_samples = vec![
        ("Hello World 123", "ASCII text to full-width"),
        ("ガ゙", "Decomposed katakana with combining mark"),
        ("檜原村", "Old kanji in place names"),
    ];

    for (text, description) in document_samples {
        let standardized = document_standardizer(text)?;
        println!("  {description}:");
        println!("    Input:  '{text}'");
        println!("    Output: '{standardized}'");
        println!();
    }

    // 3. Search index preparation
    println!("3. Search Index Preparation");
    println!("   (Use case: normalizing text for search engines)");

    let search_recipe = TransliterationRecipe {
        kanji_old_new: true,
        replace_spaces: true,
        to_halfwidth: yosina::recipes::ToHalfwidthOptions::Yes {
            hankaku_kana: false,
        },
        replace_suspicious_hyphens_to_prolonged_sound_marks: true,
        ..Default::default()
    };

    let search_normalizer = make_transliterator(&search_recipe)?;

    let search_samples = vec![
        ("東京スカイツリー", "Famous landmark name"),
        ("プログラミング言語", "Technical terms"),
        (
            "コンピューター-サイエンス",
            "Academic field with suspicious hyphen",
        ),
    ];

    for (text, description) in search_samples {
        let normalized = search_normalizer(text)?;
        println!("  {description}:");
        println!("    Original:   '{text}'");
        println!("    Normalized: '{normalized}'");
        println!();
    }

    // 4. Custom pipeline example
    println!("4. Custom Processing Pipeline");
    println!("   (Use case: step-by-step text transformation)");

    // Create multiple transliterators for pipeline processing
    let steps = vec![
        ("Spaces", vec![TransliteratorConfig::Spaces]),
        (
            "Old Kanji",
            vec![
                TransliteratorConfig::IvsSvsBase(IvsSvsBaseTransliteratorOptions {
                    mode: IvsSvsBaseMode::IvsOrSvs(false),
                    charset: Charset::Unijis2004,
                    ..Default::default()
                }),
                TransliteratorConfig::KanjiOldNew,
                TransliteratorConfig::IvsSvsBase(IvsSvsBaseTransliteratorOptions {
                    mode: IvsSvsBaseMode::Base,
                    charset: Charset::Unijis2004,
                    ..Default::default()
                }),
            ],
        ),
        (
            "Width",
            vec![TransliteratorConfig::Jisx0201AndAlike {
                fullwidth_to_halfwidth: true,
                options: Jisx0201AndAlikeTransliteratorOptions {
                    u005c_as_yen_sign: Some(false),
                    ..Default::default()
                },
            }],
        ),
        (
            "ProlongedSOundMarks",
            vec![TransliteratorConfig::ProlongedSoundMarks(Default::default())],
        ),
    ];

    let mut transliterators = Vec::new();
    for (name, config) in steps {
        let transliterator = make_transliterator(&config)?;
        transliterators.push((name, transliterator));
    }

    let pipeline_text = "hello\u{3000}world ﾃｽﾄ 檜-システム";
    let mut current_text = pipeline_text.to_string();

    println!("  Starting text: '{current_text}'");

    for (step_name, transliterator) in &transliterators {
        let previous_text = current_text.clone();
        current_text = transliterator(&current_text)?;
        if previous_text != current_text {
            println!("  After {step_name}: '{current_text}'");
        }
    }

    println!("  Final result: '{current_text}'");

    // 5. Unicode normalization showcase
    println!("\n5. Unicode Normalization Showcase");
    println!("   (Demonstrating various Unicode edge cases)");

    let unicode_recipe = TransliterationRecipe {
        replace_spaces: true,
        replace_mathematical_alphanumerics: true,
        replace_radicals: true,
        ..Default::default()
    };

    let unicode_normalizer = make_transliterator(&unicode_recipe)?;

    let unicode_samples = vec![
        (
            "\u{2003}\u{2002}\u{2000}".to_string(),
            "Various em/en spaces",
        ),
        (
            "\u{3000}\u{00A0}\u{202F}".to_string(),
            "Ideographic and narrow spaces",
        ),
        ("⺅⻌⽊".to_string(), "CJK radicals"),
        (
            "\u{1D400}\u{1D401}\u{1D402}".to_string(),
            "Mathematical bold letters",
        ),
    ];

    process_text_samples(&unicode_samples);

    for (text, description) in &unicode_samples {
        let normalized = unicode_normalizer(text)?;
        println!("  Result: '{normalized}' ({description})");
        println!();
    }

    // 6. Performance considerations
    println!("6. Performance Considerations");
    println!("   (Reusing transliterators for better performance)");

    let perf_recipe = TransliterationRecipe {
        kanji_old_new: true,
        replace_spaces: true,
        ..Default::default()
    };

    let perf_transliterator = make_transliterator(&perf_recipe)?;

    // Simulate processing multiple texts
    let texts = [
        "これはテストです",
        "檜原村は美しい",
        "hello　world",
        "プログラミング",
    ];

    println!(
        "  Processing {} texts with the same transliterator:",
        texts.len()
    );
    for (i, text) in texts.iter().enumerate() {
        let result = perf_transliterator(text)?;
        println!("    {}: '{}' → '{}'", i + 1, text, result);
    }

    println!("\n=== Advanced Examples Complete ===");

    Ok(())
}
